### PR TITLE
Online Rotations to vLLM

### PR DESCRIPTION
**Pull request by gametekker**
Papers such as QuaRot (https://arxiv.org/abs/2404.00456) have made successful attempts to greatly improve accuracy of quantized models by adding online rotations to reduce outliers.
This feature brings this capability to production vLLM, including flexible configuration for online rotations, and an optimized hadamard transformation implementation, adapted from the Tri Dao AI Lab (https://github.com/Dao-AILab/fast-hadamard-transform), optimized for ROCm.

Please refer to the design document here for full implementation detail and performance results:
https://github.com/amd-rcorzine/vllm/blob/vllm_pr_update/docs/source/features/quantization/online_rotations.md

### CODE REVIEW

File: csrc/rocm/FHT.h [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  void print() const {
+    std::cout << "batch: " << batch << std::endl;
+    std::cout << "dim: " << dim << std::endl;
+    std::cout << "log_N: " << log_N << std::endl;
+    std::cout << "x_batch_stride: " << x_batch_stride << std::endl;
+    std::cout << "out_batch_stride: " << out_batch_stride << std::endl;
+    std::cout << "scale: " << scale << std::endl;
+    std::cout << "x_ptr: " << x_ptr << std::endl;
+    std::cout << "out_ptr: " << out_ptr << std::endl;
+  }
```
[the remark this answers is not in the document]
Comment (tjtanaa) [reply csrc/rocm/FHT.h:33]: Yes. Please remove this.

File: csrc/rocm/ops.h [hunk 2]
```diff
@@ -21,4 +21,4 @@ void paged_attention(torch::Tensor& out, torch::Tensor& exp_sums,
                      int64_t block_size, int64_t max_context_len,
                      const std::optional<torch::Tensor>& alibi_slopes,
                      const std::string& kv_cache_dtype, torch::Tensor& k_scale,
-                     torch::Tensor& v_scale);
+                     torch::Tensor& v_scale);
```
Comment (tjtanaa) [inline csrc/rocm/ops.h:24]: let's revert this change.

File: docs/source/features/quantization/online_rotations.md [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+	```python
+	online_rotation_config = cast(Dict[str, Any], config.get("online_rotations"))
+	if online_rotation_config:
+	func_name,func_args=online_rotation_config['func_name'],online_rotation_config['func_args']
+	if func_name in hadamard_transform_registry:
+		func_name=hadamard_transform_registry[func_name]
+		online_rotation_method=func_name,func_args
+	else:
+		raise ValueError("hadamard rotation func_name is not found in registry")
+else:
```
Comment (tjtanaa) [inline docs/source/features/quantization/online_rotations.md:110]: There is format and rendering issue for this code block.

File: docs/source/features/quantization/online_rotations.md [hunk 4]
```diff
@@ -0,0 +1,190 @@
+# Introduction
+Recent papers QuaRot and SpinQuant papers propose a quantization method for large language models (LLMs) that use rotations to simplify quantization. It quantizes all components, including weights, activations, and KV cache, into 4 bits by removing outliers without changing the output. It accomplishes this by inserting four fundamental rotations (called R1, R2, R3 and R4 in SpinQuant), into Llama models.
+
+To support these features, the ability to add online **hadamard** rotations - explicitly rotating the input before it is passed into the quantized layer - is necessary (see figure below).
+
+![Alt text](../../assets/features/quantization/{506D28EB-F734-48C3-A3C8-0A7CF5F31141}.PNG)
+# Features
```
Comment (tjtanaa) [inline docs/source/features/quantization/online_rotations.md:7]: This file does not seem to be formatted by pre-commit. Please fix the formatting.

File: online_rotations.md [hunk 5]
```diff
@@ -0,0 +1,190 @@
+# Introduction
+Recent papers QuaRot and SpinQuant papers propose a quantization method for large language models (LLMs) that use rotations to simplify quantization. It quantizes all components, including weights, activations, and KV cache, into 4 bits by removing outliers without changing the output. It accomplishes this by inserting four fundamental rotations (called R1, R2, R3 and R4 in SpinQuant), into Llama models.
+
+To support these features, the ability to add online **hadamard** rotations - explicitly rotating the input before it is passed into the quantized layer - is necessary (see figure below).
+
+![Alt text]({506D28EB-F734-48C3-A3C8-0A7CF5F31141}.PNG)
+# Features
```
Comment (tjtanaa) [inline online_rotations.md:7]: Let's apply pre-commit hook formatting to all markdowns.

File: tests/quantization/test_online_rotation.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    isye = input_size // tp_size
+    activation_shard = activation[:, rank * (isye):(rank + 1) * (isye)]
+    activation_shard = activation_shard.contiguous().to(device)
+    rotated_activation_shard = rotation_function(activation_shard)
+    """Uniproc: Perform rotation function on entire activation - bfloat16"""
+    activation = torch.ones(2, input_size, device=device).bfloat16()
+    og_shape = activation.shape
+    X = activation.view(-1, rotation_function.chunk_size)
+    X = rotation_function.FHT(X, rotation_function.scale)
+    print(X)
```
Comment (tjtanaa) [inline tests/quantization/test_online_rotation.py:61]: Do we want to print this? if yes, can we make it more descriptive, adding a string to say we are printing X or output of FHT?

**Comment by tjtanaa** [comment]
@gametekker please add model unit tests as well.
